src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/FunctionItemCallIterator.java

```diff
@@ -113,64 +119,75 @@ private void processArguments() {
+    }
 
-            for (int i = 0; i < _functionArguments.size(); i++) {
-                argIterator = _functionArguments.get(i);
-                argName = _functionItem.getParameterNames().get(i);
-
-                if (argIterator == null) { // == ArgumentPlaceholder
-                    partialAppParamNames.add(argName);
-                    partialAppParamTypes.add(_functionItem.getSignature().getParameterTypes().get(i));
-                } else {
-                    List<Item> argValue = getItemsFromIteratorWithCurrentContext(argIterator);
-                    argumentValues.put(argName, argValue);
-                }
-            }
+    private DynamicContext createNewDynamicContextWithArguments(DynamicContext context) {
+        this.validateAndReadArguments();
 
-            // partial application should return a new FunctionItem with given parameters set as NonLocalVariables
-            // and argument placeholders as new parameters to the new FunctionItem
-            FunctionItem partiallyAppliedFunction = new FunctionItem(
-                    new FunctionIdentifier("", partialAppParamNames.size()),
-                    partialAppParamNames,
-                    new FunctionSignature(
-                            partialAppParamTypes,
-                            _functionItem.getSignature().getReturnType()
-                    ),
-                    _functionItem.getBodyIterator(),
-                    argumentValues
-            );
-            _functionBodyIterator = new FunctionRuntimeIterator(partiallyAppliedFunction, getMetadata());
-        } else {
-            // calculate argument values
-            for (int i = 0; i < _functionArguments.size(); i++) {
-                argIterator = _functionArguments.get(i);
-                argName = _functionItem.getParameterNames().get(i);
+        RuntimeIterator argIterator;
+        String argName;
 
-                List<Item> argValue = getItemsFromIteratorWithCurrentContext(argIterator);
-                argumentValues.put(argName, argValue);
-            }
+        // calculate argument values
+        Map<String, List<Item>> argumentValues = new LinkedHashMap<>(_functionItem.getNonLocalVariableBindings());
+        for (int i = 0; i < _functionArguments.size(); i++) {
+            argIterator = _functionArguments.get(i);
+            argName = _functionItem.getParameterNames().get(i);
 
-            // place argument values into dynamic context
-            _currentDynamicContext = new DynamicContext(_currentDynamicContext);
-            for (Map.Entry<String, List<Item>> argumentEntry : argumentValues.entrySet()) {
-                _currentDynamicContext.addVariableValue(
-                    argumentEntry.getKey(),
-                    argumentEntry.getValue()
-                );
-            }
-            _functionBodyIterator = _functionItem.getBodyIterator();
+            List<Item> argValue = getItemsFromIteratorWithCurrentContext(argIterator);
```

Comment:
shouldn't context (the parameter) be used as well here?

Reply:
I was investigating this as well, as it seemed like a potential bug that required testing. However, I just noticed that if the function is partially executed, the execution mode is set to Local. Hence no problems came up previously. I will try to clarify this in the code by adding throwing an exception if execution reaches RDD evaluation during a partial application.

Reply:
Could you take a look at the latest commit now

Reply:
I did and it looks good. I still have a concern with the line above: 

List<Item> argValue = getItemsFromIteratorWithCurrentContext(argIterator);

Indeed, this method createNewDynamicContextWithArguments is also called in getRDDAux, so that it might make more sense to past the dynamic context supplied as a parameter (context).

A query like this might be a way to test:

declare function f($i) { parallelize($i) }; let $a := 3 return f($a)